### PR TITLE
Adds Whims (AKA ain't nothin in the contributing guidelines says you can't make a dog play basketball)

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
Recently, I watched that one clip from Aid Bud and wanted to make Ian play basketball, so after writing a crappy draft for the behavior directly in the Life() proc for dogs, I sat down for a few days and hashed out a datumized AI system for simple_animals with hotswappable behaviors. These are called Whims.

Basically these datums run scans every few ticks to see if their conditions to activate are met, and if they are, the Whim steps up and takes control of the mob for a bit until it decides it's done or gets frustrated (generally these last no more than 15-20 seconds in total). Only one whim can be in control at a time, and generally you have to wait for the current one to be finished before the others can resume scanning. If no Whim is in control, the mob will act as simple_animals normally do, I.E. brainlessly walking around occasionally emoting.

Whims are intended to be simple behaviors that are triggered by the mob's direct environment. Here's a list of a few of the behaviors I've written:

- Not only will cats hunt down and kill nearby mice, if they do so near someone, they'll be a nice little friend and bring the dead mouse to the person's feet as a gift! Cute! Gross!
- In addition to eating any and all random foods left on the ground, dogs will sometimes run off with dismembered limbs they find to gnaw on the bones. Letting a pack of dogs loose in a busy medbay has never been so much fun!
- Cats left near open bottles and reagent containers will sometimes knock them over because they're jerks, causing them to spill. Don't let Runtime into the Pharmacy!
- Baby making has been made a Whim as well instead of existing on the base simple animal class and only being used for a handful of mobs.
- I reimplemented my previous catnip PR, #52866, and it's much less horrible and much more modular now. Slip an open can of cat food into someone's pocket and watch them get chased by a swarm of cats! Hilarious!
- Ian will become aggressive if people who aren't the HoP or the Captain enter his owner's office. His outbursts are normally limited to growling and barking, but if you're standing next to the disposal bin, he might just push you in and flush you! If you're the HoP and need to invite someone into your office, make sure you put Ian in his cage!
- Probably my favorite, Ian will also mourn the death of his owner. Upon seeing the corpse of the HoP, Ian will hover around it, nuzzling and pawing at the body trying to wake them up. Even more impressive, if the body is hidden in the dark and light is nearby, Ian will drag the body towards better lit tiles to help save you! What a good boy!
- And, of course, hold a basketball around a dog at your own peril.

The neat thing is that, like components, these behaviors can be very generalized and applied to different simple_animals easily. The only thing tying these to specific mobs/targets are the descriptions for visible messages of them barking or meowing or whatever, and those are easy enough to modularize.

### Performance Considerations
Full disclosure, I didn't really plan any of this out. This entire system grew organically out of me wanting to add little funny behaviors to simple animals, with structure being added where I needed it along the way. I tried anticipating the added costs by having niche/expensive scans run less frequently (say, every 10-12 ticks instead of 2-3), but I'd like to hear thoughts on if this is actually viable and any possible improvements. And, of course, some kind of safeguard will need to be put in for petsplosions so we aren't juggling a bunch of checks for 200 corgis.

## Why It's Good For The Game
There've been other attempts to improve the AI in SS13 with robust systems like GOAP, but this takes a much simpler approach. When people think of improving AI in SS13 they immediately think of making complex, human-like bots that can trick you into thinking they're people but, like... the entire cool thing about SS13 is that you're interacting with a bunch of other people, so why would we dilute that with fake non-people? That's why I boldly decided to dedicate my efforts to making animals do funny things instead, and make doing that easy.

## Changelog
:cl: Ryll/Shaps
add: Cats, dogs, and Ians of the station have received some new behaviors! Go play basketball on the holodeck with Ian sometime!
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
I dig the idea, and ther technical concept behind this isn't bad, but I think we're better off doing a more full-fledged AI refactor to add something like this, as this will most likely add more technical debt to simple mobs.

**Comment by Qustinnus** [comment]
> There've been other attempts to improve the AI in SS13 with robust systems like GOAP, but this takes a much simpler approach. When people think of improving AI in SS13 they immediately think of making complex, human-like bots that can trick you into thinking they're people but, like... the entire cool thing about SS13 is that you're interacting with a bunch of other people, so why would we dilute that with fake non-people? That's why I boldly decided to dedicate my efforts to making animals do funny things instead, and make doing that easy.

in response to this, I agree that systems like GOAP are more than we need, but that does not mean adding an actual system for AI isn't the right solution. We need to work on something like behavior trees where we have a uniform system for AI behavior implementation that is modular and where we can re-use behavior. It's not about "What is it capable of" but more about "How modular / scalable is it". In that sense, you have the right idea with something like this but it's too specialized and doesn't solve the actual issue with mob AI; just solves the need you specifically have while leaving the existing issues.

**Comment by Qustinnus** [comment]
if you're interested in hearing more about this you can also DM me on discord, I have a mental image of how an actual AI re-work would work, but I'm currently working on 2 big systems so I'm not planning to work on it until theyre done.

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -46,6 +46,8 @@ GLOBAL_LIST_INIT(turfs_without_ground, typecacheof(list(
 //Mobs
 #define isliving(A) (istype(A, /mob/living))
 
+#define issimp(A) (istype(A, /mob/living/simple_animal))
```
Comment (Jared-Fogle) [inline code/__DEFINES/is_helpers.dm:49]: don't make me be the killer of good times shaps

**Comment by Ryll-Ryll** [comment]
Why would I be interested in throwing out all of these improvements, which have demonstrably helped me create content I wouldn't be able to otherwise and has helped clean up several mobs' Life() procs, for some other unseen plan that exists only as a mental image in the mind of a person who admits they have no interest in tangibly committing to it right now

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (Ryll-Ryll) [reply code/__DEFINES/is_helpers.dm:49]: man i thought that'd last longer than 11 minutes

Comment (Jared-Fogle) [reply code/__DEFINES/is_helpers.dm:49]: i'm sorry shaps

**Comment by Qustinnus** [comment]
no need to be so hostile, I'm just sharing what I think of the topic and figured you might be interested.

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/__DEFINES/is_helpers.dm:49]: this must stay in the codebase

**Comment by Ryll-Ryll** [comment]
Sorry, I didn't mean for that to be hostile, I'm just a bit tired and probably wrote that a bit more matter-of-factly than I needed to. I took a few courses in AI during my time in college and have definitely had ideas about implementing neat, sophisticated systems in SS13, but the truth is that I don't believe the game is a very good use-case for advanced AI systems. Besides the massive complexity of the world's state, players have already shown they're more than happy to treat a little floorbot as special just because the game says it has a knife taped to it and causes people to fall over when they cross it, which is a pretty amazing return on investment for something that took a day or two to create. My goal here is to make those scenarios easier to create, which I think this does well

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/__DEFINES/is_helpers.dm:49]: @Jared-Fogle ok boomer

**Comment by wesoda25** [comment]
Neat PR shaps

**Comment by Ryll-Ryll** [comment]
After some more review, I think that there are better ways to accomplish my goals here that don't involve such finnicky structure, which I'll look into after the freeze

**Comment by LemonInTheDark** [comment]
😭

**Pull request closed without merging**